src/kv/RocksDBStore.cc

```diff
@@ -2729,16 +2741,25 @@ class ShardMergeIteratorImpl : public KeyValueDB::IteratorImpl {
   const RocksDBStore* db;
   KeyLess keyless;
   string prefix;
+  const KeyValueDB::IteratorBounds bounds;
   std::vector<rocksdb::Iterator*> iters;
 public:
   explicit ShardMergeIteratorImpl(const RocksDBStore* db,
 				  const std::string& prefix,
-				  const std::vector<rocksdb::ColumnFamilyHandle*>& shards)
-    : db(db), keyless(db->comparator), prefix(prefix)
+				  const std::vector<rocksdb::ColumnFamilyHandle*>& shards,
+                  const KeyValueDB::IteratorBounds bounds)
+    : db(db), keyless(db->comparator), prefix(prefix), bounds(bounds)
   {
     iters.reserve(shards.size());
     for (auto& s : shards) {
-      iters.push_back(db->db->NewIterator(rocksdb::ReadOptions(), s));
+      auto options = rocksdb::ReadOptions();
+      if (bounds.upper_bound) {
+        options.iterate_upper_bound = new rocksdb::Slice(*this->bounds.upper_bound);
```

Comment:
does rocksdb free this memory? i don't think it does, so we need to manage the memory ourselves

i pushed a commit https://github.com/cbodley/ceph/commit/9baa3db8daca52a95b0d817db7de489e3c53e4ad as an example (based on your earlier [wip-cfsnyder-rgw-updates-pacific ](https://github.com/ceph/ceph-ci/commits/wip-cfsnyder-rgw-updates-pacific) branch), that constructs these slices as member variables instead

Reply:
I had originally stored pointers as member variables and was deleting them in the destructor, but was getting runtime errors from the allocator. I had assumed at the time that rocksdb was freeing this memory and the allocator was complaining due to a double free - but it wasn't clear. I'm testing your branch on my test cluster now and will definitely update to use your approach if things behave properly.

Reply:
we chatted a bit about this on irc, and agreed that the `.value_or("")` part there is wrong because it returns a temporary std::string. i think we'd need a helper function like this:
```c++
static rocksdb::Slice make_slice(const std::optional<std::string>& bound) {
  if (bound) {
    return {*bound};
  } else {
    return {};
  }
}
```
so we can construct the member Slices like:
```c++
      bounds(std::move(bounds_)),
      iterate_lower_bound(make_slice(bounds.lower_bound)),
      iterate_upper_bound(make_slice(bounds.upper_bound))
```